### ISSUE TITLE
WebApp not starting - when connection string is bad

### ISSUE THREAD

**Issue by bobbyangers** [dj-nitehawk/MongoDB.Entities#149]
When there happens to have an error in the connection string, for example a password rotation/change.   The WebApp won't even start.

``System.AggregateException: One or more errors occurred. (Unable to authenticate using sasl protocol mechanism SCRAM-SHA-1.)
 ---&gt; MongoDB.Driver.MongoAuthenticationException: Unable to authenticate using sasl protocol mechanism SCRAM-SHA-1.
 ---&gt; MongoDB.Driver.MongoCommandException: Command saslContinue failed: Authentication failed..``

This happen because of this....
[DB.Init](https://github.com/dj-nitehawk/MongoDB.Entities/blob/15ba18486ea6bbe58f62573f59157e7f7fe6baf6/MongoDB.Entities/DB/DB.cs#L81)

It would be to if the parameter => ``bool skipNetworkPing = false`` was exposed to optionally do the ping.

**Comment by dj-nitehawk** [comment]
will expose the parameter eventually. probably after ahmednfwela's PR gets merged.

however, should a webapp even start when it can't connect to the database?
isn't it better to usually log/notify admin within a try/catch when the db can't be accessed during startup?

btw, `DBContext` [doesn't attempt to connect](https://mongodb-entities.com/wiki/DB-Instances.html#create-an-instance) on init.

**Comment by bobbyangers** [comment]
@dj-nitehawk yes, it should start.  When in the cloud.. app fail to start is the worst to debug....  if you are lucky, you have a log without a stack trace....  especially in PROD where it's compiled in Release.

much easier to let the app start, and have a health page that checks the connection (ie ping).  For the moment, I put the DB.Init in a try/catch block.

**Comment by paolobriones** [comment]
I was actually going to open a ticket similar to this @bobbyangers and @dj-nitehawk 

What I am proposing is to have a dedicated method in the API where we can pass in a connection string and have the API validate it and return a bool if the API is able to connect to the MongoDB server.

My use-case is that when my application runs for the first time after installation, the user is greeted with a setup screen where I ask the user for MongoDB credentials and server address.

This might also be a reasonable sanity check during the startup of the application instead of relying on a thrown exception to know that the connection to MongoDB is invalid.

**Comment by paolobriones** [comment]
Just thinking about it, there should be 2 sanity check methods:

**by raw connection string** 
verify if the connection string is actually valid and the credentials/address are all correct

**by DBContext**
Just be able to ping and verify if the initialized DBContext is still able to connect to MongoDB. Just to verify if MongoDB is still alive

**Comment by paolobriones** [comment]
At the moment, i am currently doing this to test the connection string:

``` csharp
public static bool TestConnection(string databasename, string connectionstring) {
	GuardAgainst.ArgumentBeingNullOrEmpty(databasename);
	GuardAgainst.ArgumentBeingNullOrEmpty(connectionstring);

	MongoClient client = new(connectionstring);
	IMongoDatabase database = client.GetDatabase(databasename);

	return database.RunCommandAsync((Command<BsonDocument>)"{ping:1}").Wait(1000);
}
```

**Comment by dj-nitehawk** [comment]
@paolobriones have you seen [this method](https://mongodb-entities.com/wiki/Multiple-Databases.html#check-if-a-database-is-still-accessible )?

that method [does the ping internally](https://github.com/dj-nitehawk/MongoDB.Entities/blob/15ba18486ea6bbe58f62573f59157e7f7fe6baf6/MongoDB.Entities/Extensions/Database.cs#L26-L40).

anyhow, will look into this more once pr #125 is done.

**Comment by paolobriones** [comment]
@dj-nitehawk I did not see those!! I will take a look at them this weekend

**Comment by paolobriones** [comment]
@dj-nitehawk The utility methods look good!

Can we move those into their own page in documentation, I completely missed out on those

**Comment by dj-nitehawk** [comment]
@paolobriones 
> Can we move those into their own page in documentation

done!

**Comment by bobbyangers** [comment]
@dj-nitehawk where are we at with this ?

I have an Azure Function, that gets "unloaded" (ie Serverless);   

I don't need it to check the connection everytime it "wakes up".   

Could we not skip the check base on an Environment variable ?  `MONGO_ENTITIES_SKIP_PING`

**Comment by dj-nitehawk** [comment]
@ahmednfwela is #125 abandoned?

**Comment by ahmednfwela** [comment]
nope, I just didn't have enough time until lately, so I am working on it again

**Comment by dj-nitehawk** [comment]
great!

**Comment by dj-nitehawk** [comment]
@ahmednfwela great!

@bobbyangers you can use [DBContext](https://mongodb-entities.com/wiki/DB-Instances.html#create-an-instance) for now, which doesn't do the ping.
